doc/mkencap.1
.TH mkencap 1 "Oct 2002" "University of Illinois" "Package Management"
.SH NAME
mkencap \- Encap package creation tool
.SH SYNOPSIS
\fBmkencap [options] -b [-DUPCBIT] \\
.br
        [-m m4_cmd] [-M m4_outfile] \\
.br
        [-d download_dir] [-F environment_file] \\
.br
        [-O build_tree] [-S src_tree] profile

mkencap [options] \\
.br
        [ -e [-n] [-a "encapinfo_directive ..."] ] \\
.br
        [ -c [-o outfile] ] pkgspec\fP
.SH VERSION
This man page documents version 2.3 of \fBmkencap\fP.
.SH DESCRIPTION
\fBmkencap\fP is a tool to help create \fIEncap\fP 2.1 packages.  It can
create package directories, generate \fIencapinfo\fP files, and create
package archives.
.SH MODES
Any combination of these modes may be specified when \fBmkencap\fP is
invoked.  Directory initialization is done first, if selected.  Then,
if the \fIencapinfo\fP file is needed, it is created.  Finally, a tar
archive is created if desired.
.TP
.B -b
Build the package from a profile.  If the \fI-b\fP option is specified,
\fIpackage\fP should be the path to the file containing the package's
profile.  (See \fBPACKAGE PROFILE OPTIONS\fP below for other related
options.)
.TP
.B -e
Create an \fIencapinfo\fP file in the package directory.
(See \fBENCAPINFO OPTIONS\fP below for other related options.)
.TP
.B -c
Create archive.  \fBmkencap\fP will create a (optionally compressed)
tar archive of \fIpackage\fP.  (See \fBARCHIVE CREATION OPTIONS\fP
below for other related options.)
.PP
If none of these options are selected, \fBmkencap\fP will create an
\fIencapinfo\fP file and a tar archive.
.SH GENERAL OPTIONS
These options may be used in combination with any other options:
.TP
.B -E encap_format
Specify the Encap package format version of the new package.  The
default is \fI2.1\fP.
.TP
.B -p platform
Set the package's Encap platform name to \fIplatform\fP.  If this option
is not specified, the platform name is set to that of the host that
\fBmkencap\fP is running on.
.TP
.B -A platform_suffix
Appends \fIplatform_suffix\fP to the package's Encap platform name.
.TP
.B -f
Force \fBmkencap\fP to overwrite an existing \fIencapinfo\fP file.
.TP
.B -q
Quiet mode.  Only error messages are output to stdout.
.TP
.B -s source_dir
Tells \fBmkencap\fP to use \fIsource_dir\fP as the \fIEncap\fP source
directory.  If the source is not specified but the target is,
\fBmkencap\fP will assume the relative path from the target to the
source is \fIencap/\fP.  If neither is specified, \fBmkencap\fP will check the
contents of the \fPENCAP_SOURCE\fP environment variable.  If that is
not set, it will use the default set at compile-time (usually
\fI/usr/local/encap\fP).
.TP
.B -t target_dir
Tells \fBmkencap\fP to use \fItarget_dir\fP as the \fIEncap\fP target
directory.  If the target is not specified but the source is,
\fBmkencap\fP will assume the relative path from the source to the
target is \fI../\fP.  If neither is specified, \fBmkencap\fP will check the
contents of the \fBENCAP_TARGET\fP environment variable.  If that is
not set, it will use the default set at compile-time (usually
\fI/usr/local\fP).
.TP
.B -v
Increase verbosity.  This option can be specified multiple times.
.TP
.B -V
Print \fBmkencap\fP version and compile-time defaults.
.SH PACKAGE PROFILE OPTIONS
These options are only meaningful when processing package profiles (see
\fI-b\fP option above):
.TP
.B -D
Download source distribution if not already present.  The downloaded
file is placed in the common source tree.
.TP
.B -U
Unpack source distribution.
.TP
.B -P
Apply any necessary patches to source distribution.
.TP
.B -C
Configure source distribution.
.TP
.B -B
Build source distribution.
.TP
.B -I
Install source distribution to create an Encap package.
(This usually needs to be done as \fIroot\fP.)
.TP
.B -T
Tidy source distribution.
.PP
If none of the \fI-DUPCBIT\fP options are selected, the default
is \fI-DUPCB\fP.
.TP
.B -m m4_cmd
Use the command \fIm4_cmd\fP to preprocess the package profile.
If the \fI-m\fP option is not specified, \fBmkencap\fP will check the
\fIMKENCAP_M4_COMMAND\fP environment variable.  If that variable is
unset, \fBmkencap\fP will use the string "m4", which means that it will
use the first \fBm4\fP binary available in the \fIPATH\fP.

.B Note:
The preprocessing command must point to GNU \fBm4\fP.  If the
first \fBm4\fP binary in your \fIPATH\fP is not GNU \fBm4\fP, you must
use either the \fIMKENCAP_M4_COMMAND\fP environment variable or the
\fI-m\fP option to specify the path to GNU \fBm4\fP.

This option can be used to simply override which \fBm4\fP binary is
used, as in:

  mkencap -m /usr/local/bin/m4 -b pkgspec.ep

However, it can also be used to add options to be passed to the \fBm4\fP
command, as in:

  mkencap -m "m4 -Dmacro=value" -b pkgspec.ep
.TP
.B -d download_dir
Tells \fBmkencap\fP where to put source files as they are downloaded.
If the \fI-d\fP option is not specified, \fBmkencap\fP will check the
\fIMKENCAP_DOWNLOAD_DIR\fP environment variable.  If that variable is
unset, \fBmkencap\fP will use the default set at compile-time (usually
\fI/usr/local/src\fP).
.TP
.B -F environment_file
Tells \fBmkencap\fP to read environment variable settings from
\fIenvironment_file\fP.  If the \fI-F\fP option is not specified,
\fBmkencap\fP will use the default set at compile-time (usually
\fI/usr/local/etc/mkencap_environment\fP).
.TP
.B -S src_tree
Tells \fBmkencap\fP where the common source tree is located.  For
packages that can be built in a seperate directory from where the source
code is located, the source code will be stored in the common source
tree.  If the \fI-S\fP option is not specified, \fBmkencap\fP will check
the \fIMKENCAP_SRC_TREE\fP environment variable.  If that variable is
unset, \fBmkencap\fP will store the source code in the build tree.
.TP
.B -O build_tree
Tells \fBmkencap\fP where the build tree is located.  The build tree is
used for building packages from source code.  If the \fI-O\fP option is
not specified, \fBmkencap\fP will check the \fIMKENCAP_BUILD_TREE\fP
environment variable.  If that variable is unset, \fBmkencap\fP will
use the default set at compile-time (usually \fI/usr/local/src\fP).

If the build tree path contains the string "%p", it will be replaced
with the platform name.
.TP
.B -M m4_outfile
Tells \fBmkencap\fP to write the output of \fBm4\fP to the file
\fIm4_outfile\fP.  If \fIm4_outfile\fP is "-", the output will be
written to \fIstdout\fP.

This option is useful for debugging problems caused by \fBm4\fP
preprocessing.
.SH ENCAPINFO OPTIONS
These options are only meaningful when creating an \fIencapinfo\fP file
(see \fI-e\fP option above):
.TP
.B -n
Do not add default \fIcontact\fP, \fIdate\fP, and \fIplatform\fP fields to
the \fIencapinfo\fP file.

When adding the \fIcontact\fP field, \fBmkencap\fP will first check the
environment variable \fIENCAP_CONTACT\fP.  If unset, it will use the
username of the invoking user and the hostname of the host it was run on.
.TP
.B -a "encapinfo_directive ..."
Adds the specified directive to the \fIencapinfo\fP file.  Syntax checking
is done to ensure that the line is parsable.
.SH ARCHIVE CREATION OPTIONS
This option is only meaningful when creating an archive file (see
\fI-c\fP option above):
.TP
.B -o outfile
Specify the output filename when creating an archive.  If the filename ends
in \fI.tgz\fP or \fI.gz\fP, it is compressed using \fBgzip\fP(1).  If the
filename ends in \fI.Z\fP, it is compressed using \fBcompress\fP(1).  If
the filename ends in \fI.bz2\fP, it is compressed using \fBbzip2\fP(1).

For Encap 2.0 packages, the default is \fIpkgspec.tar.gz\fP.

For Encap 2.1 packages, the default is
\fIpkgspec-encap-platform.tar.gz\fP.  The \fIplatform\fP string will be
read from the \fIencapinfo\fP file.
.SH EXAMPLES
To create an \fIencapinfo\fP file for the package "pkg-2.1" with the
directives "linkname bin/foo quux" and "linkname man/man1/foo.1 quux.1":

.RS
mkencap -ea "linkname bin/foo quux" \\
.br
        -a "linkname man/man1/foo.1 quux.1" pkg-2.1
.RE

To create an \fIencapinfo\fP file for the package "pkg-3.0" containing
the directives "exclude bin/burfle" and "linkdir lib/pkg", and create
an archive of the package in your home directory:

.RS
mkencap -a "exclude bin/burfle" -a "linkname lib/pkg" \\
.br
        -o ~/pkg-3.0-encap-share.tgz pkg-3.0
.RE
.SH ENVIRONMENT VARIABLES
.B ENCAP_SOURCE
.br
.B ENCAP_TARGET
.br
.B ENCAP_CONTACT
.br
.B MKENCAP_DOWNLOAD_DIR
.br
.B MKENCAP_SRC_TREE
.br
.B MKENCAP_BUILD_TREE
.br
.B MKENCAP_M4_COMMAND
.SH FILES
.I /usr/local/encap/pkgname/encapinfo
.br
.I /usr/local/etc/mkencap_environment
.SH SEE ALSO
.BR epkg (1),
.BR encap_profile (5)
.br
The Encap Archive (\fIhttp://www.encap.org/\fP)
.SH AUTHOR
Mark D. Roth <roth@feep.net>


NEWS
User-visible changes from 2.2.x to 2.3.x:
-----------------------------------------

* added HTTP download support using cURL (see README file for details)

* added automatic reconnection if FTP server times out between
  downloads; use "-T" option to disable

* new package profile format used by mkencap for automated package creation
  (see encap_profile(5) man page for explanation of the profile format)
  (requires libexpat to be installed at build time and GNU m4 to be
  installed at run-time; see README for details)

* removed mkencap directory creation mode ("mkencap -i")

* many changes to mkencap command line interface

* added encapinfo_parse_directive() to libencap API

* changed Encap 2.1 platform name for HP-UX from "pa1.1-hpux11.00"
  to "pa-hpux11.00"
  (encap_platform_compat() will still recognize the old name)



mkencap/mkencap.c
/*
**  Copyright 1998-2003 University of Illinois Board of Trustees
**  Copyright 1998-2004 Mark D. Roth
**  All rights reserved.
**
**  mkencap.c - encap package creation tool
**
**  Mark D. Roth <roth@feep.net>
*/

#include <mkencap.h>

#include <stdio.h>
#include <errno.h>
#include <sys/param.h>
#include <time.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>

#ifdef STDC_HEADERS
# include <string.h>
# include <stdlib.h>
#endif

#ifdef HAVE_UNISTD_H
# include <unistd.h>
#endif

#ifdef HAVE_LIBTAR
# include <libtar.h>
#endif

#ifdef HAVE_LIBZ
# include <zlib.h>
#endif

#ifdef HAVE_LIBEXPAT
# include <expat.h>
#endif

#ifdef HAVE_LIBFGET
# include <libfget.h>
#endif

#ifdef HAVE_LIBCURL
# include <curl/curl.h>
#endif


/* modes */
#define MODE_BUILD		1
#define MODE_CREATE		2
#define MODE_ENCAPINFO		4


/* global settings */
char platform[MAXPATHLEN] = "";
short verbose = 1;
unsigned long mkencap_opts = 0;
encapinfo_t pkginfo;
char common_src_tree[MAXPATHLEN] = DEFAULT_COMMON_SRC_TREE;
char build_tree[MAXPATHLEN] = DEFAULT_BUILD_TREE;
char download_tree[MAXPATHLEN] = DEFAULT_DOWNLOAD_DIR;


static void
usage(void)
{
	fprintf(stderr, "Options: [-qvVf] [-s source] [-t target]\n");
	fprintf(stderr, "         [-E encap_format] [-p platform] "
		"[-A platform_suffix]\n");
	fprintf(stderr, "Usage:\n");
#ifdef HAVE_LIBEXPAT
	fprintf(stderr, "   mkencap [options] -b [-DUPCBIT] [-m m4_cmd] "
		"[-M m4_outfile] \\\n");
	fprintf(stderr, "           [-d download_dir] "
		"[-F environment_file] \\\n");
	fprintf(stderr, "           [-O build_tree] [-S src_tree] profile\n");
#endif
	fprintf(stderr, "   mkencap [options] \\\n");
	fprintf(stderr, "           [ -e [-n] "
		"[-a \"encapinfo_directive ...\"] ] \\\n");
	fprintf(stderr, "           [ -c [-o outfile] ] pkgspec\n");
	exit(1);
}


static void
print_defaults(void)
{
#ifdef HAVE_LIBEXPAT
	XML_Expat_Version expat_ver;
#endif

	printf("mkencap %s by Mark D. Roth <roth@feep.net>\n", PACKAGE_VERSION);
	printf("   Build Platform:\t\t%s\n", PLATFORM);
#ifdef HAVE_LIBTAR
	printf("   Builtin tar:\t\t\tlibtar-%s\n", libtar_version);
#else
	printf("   Builtin tar:\t\t\tN/A\n");
#endif
#ifdef HAVE_LIBZ
	printf("   Builtin gzip:\t\tzlib-%s\n", ZLIB_VERSION);
#else
	printf("   Builtin gzip:\t\tN/A\n");
#endif
#ifdef HAVE_LIBEXPAT
	expat_ver = XML_ExpatVersionInfo();
	printf("   Builtin XML:\t\t\texpat-%d.%d.%d\n",
	       expat_ver.major,
	       expat_ver.minor,
	       expat_ver.micro);
#else
	printf("   Builtin XML:\t\t\tN/A\n");
#endif
#ifdef HAVE_LIBFGET
	printf("   Builtin FTP:\t\t\tlibfget-%s\n", libfget_version);
#else
	printf("   Builtin FTP:\t\t\tN/A\n");
#endif
#ifdef HAVE_LIBCURL
	printf("   Builtin HTTP:\t\tlibcurl-%s\n", LIBCURL_VERSION);
#else
	printf("   Builtin HTTP:\t\tN/A\n");
#endif
	printf("   Default Encap Source:\t%s\n", DEFAULT_SOURCE);
	printf("   Default Encap Target:\t%s\n", DEFAULT_TARGET);
	printf("   Default Download Dir:\t%s\n", DEFAULT_DOWNLOAD_DIR);
	printf("   Default Build Tree:\t\t%s\n", DEFAULT_BUILD_TREE);
	printf("   Default Common Source Tree:\t%s\n",
	       strlen(DEFAULT_COMMON_SRC_TREE) > 0
	       ? DEFAULT_COMMON_SRC_TREE
	       : "N/A");
	exit(0);
}


#ifdef HAVE_LIBEXPAT

static void
mkencap_init_paths(char *cwd, char *opt_download,
		   char *opt_common_src, char *opt_build)
{
	char buf[MAXPATHLEN];
	char *cp;

	/* initialize download dir */
	if (opt_download == NULL)
	{
		cp = getenv("MKENCAP_DOWNLOAD_DIR");
		if (cp != NULL)
			strlcpy(download_tree, cp, sizeof(download_tree));
	}
	else if (opt_download[0] != '/')
	{
		snprintf(buf, sizeof(buf), "%s/%s", cwd, opt_download);
		encap_cleanpath(buf, download_tree, sizeof(download_tree));
	}
	else
		strlcpy(download_tree, opt_download, sizeof(download_tree));

	/* initialize common src dir */
	if (opt_common_src == NULL)
	{
		cp = getenv("MKENCAP_SRC_TREE");
		if (cp != NULL)
			strlcpy(common_src_tree, cp, sizeof(common_src_tree));
	}
	else if (opt_common_src[0] != '/')
	{
		snprintf(buf, sizeof(buf), "%s/%s", cwd, opt_common_src);
		encap_cleanpath(buf, common_src_tree, sizeof(common_src_tree));
	}
	else
		strlcpy(common_src_tree, opt_common_src,
			sizeof(common_src_tree));

	/* initialize build dir */
	if (opt_build == NULL)
	{
		cp = getenv("MKENCAP_BUILD_TREE");
		if (cp != NULL)
			strlcpy(build_tree, cp, sizeof(build_tree));
	}
	else if (opt_build[0] != '/')
	{
		snprintf(buf, sizeof(buf), "%s/%s", cwd, opt_build);
		encap_cleanpath(buf, build_tree, sizeof(build_tree));
	}
	else
		strlcpy(build_tree, opt_build, sizeof(build_tree));

	/* replace "%p" with platform name in build_tree */
	strlcpy(buf, build_tree, sizeof(buf));
	encap_gsub(buf, "%p", platform, build_tree, sizeof(build_tree));
}

#endif /* HAVE_LIBEXPAT */


static int
mkencap_print_func(ENCAP *e, encap_source_info_t *srcinfo,
		   encap_target_info_t *tgtinfo, unsigned int type,
		   char *fmt, ...)
{
	va_list args;

	printf("mkencap: ");
	va_start(args, fmt);
	vprintf(fmt, args);
	va_end(args);

	return 0;
}


/* main() */
int
main(int argc, char *argv[])
{
	int i, exitval = 0;
	char pkgdir[MAXPATHLEN], cwd[MAXPATHLEN];
	char outfile[MAXPATHLEN] = "";
	char errbuf[1024];
	char *platform_suffix = NULL;
	char *opt_src = NULL, *opt_tgt = NULL;
	unsigned short mode = 0;
	char *pkgspec;
	ENCAP *e = NULL;
	encapinfo_t *eip;
#ifdef HAVE_LIBEXPAT
	char profile[MAXPATHLEN];
	unsigned long profile_modes = 0;
	encap_profile_t *epp;
	char *m4_cmd;
	char *opt_env_file = NULL, *opt_m4_outfile = NULL;
	char *opt_common_src = NULL, *opt_build = NULL, *opt_download = NULL;
#endif /* HAVE_LIBEXPAT */

	/* initialize encapinfo structure */
	encapinfo_init(&pkginfo);
	pkginfo.ei_pkgfmt = strdup("2.1");

#ifdef HAVE_LIBEXPAT
	/* initialize m4 command */
	m4_cmd = getenv("MKENCAP_M4_COMMAND");
	if (m4_cmd == NULL)
		m4_cmd = "m4";
#endif /* HAVE_LIBEXPAT */

	/* parse arguments */
	while ((i = getopt(argc, argv,
			   "a:A:bBcCd:DeE:fF:hIm:M:"
			   "no:O:p:Pqs:S:t:TUvV")) != -1)
	{
		switch (i)
		{
		case 'V':
			print_defaults();
			break;
		case 'v':
			verbose++;
			break;
		case 'q':
			verbose = 0;
			break;
		case 's':
			opt_src = optarg;
			break;
		case 't':
			opt_tgt = optarg;
			break;
		case 'f':
			BIT_TOGGLE(mkencap_opts, MKENCAP_OPT_OVERWRITE);
			break;
		case 'E':
			free(pkginfo.ei_pkgfmt);
			pkginfo.ei_pkgfmt = strdup(optarg);
			break;
		case 'p':
			strlcpy(platform, optarg, sizeof(platform));
			break;
		case 'A':
			platform_suffix = optarg;
			break;
		case 'c':
			BIT_SET(mode, MODE_CREATE);
			break;
		case 'e':
			BIT_SET(mode, MODE_ENCAPINFO);
			break;
#ifdef HAVE_LIBEXPAT
		case 'b':
			BIT_SET(mode, MODE_BUILD);
			break;
		case 'D':
			BIT_TOGGLE(profile_modes,
				   MKENCAP_PROFILE_DOWNLOAD);
			continue;
		case 'U':
			BIT_TOGGLE(profile_modes,
				   MKENCAP_PROFILE_UNPACK);
			continue;
		case 'P':
			BIT_TOGGLE(profile_modes,
				   MKENCAP_PROFILE_PATCH);
			continue;
		case 'C':
			BIT_TOGGLE(profile_modes,
				   MKENCAP_PROFILE_CONFIGURE);
			continue;
		case 'B':
			BIT_TOGGLE(profile_modes,
				   MKENCAP_PROFILE_BUILD);
			continue;
		case 'I':
			BIT_TOGGLE(profile_modes,
				   MKENCAP_PROFILE_INSTALL);
			continue;
		case 'T':
			BIT_TOGGLE(profile_modes,
				   MKENCAP_PROFILE_TIDY);
			continue;
		case 'O':
			opt_build = optarg;
			break;
		case 'S':
			opt_common_src = optarg;
			break;
		case 'm':
			m4_cmd = optarg;
			break;
		case 'd':
			opt_download = optarg;
			break;
		case 'M':
			opt_m4_outfile = optarg;
			break;
		case 'F':
			opt_env_file = optarg;
			break;
#endif /* HAVE_LIBEXPAT */
		case 'o':
			strlcpy(outfile, optarg, sizeof(outfile));
			break;
		case 'n':
			BIT_TOGGLE(mkencap_opts, MKENCAP_OPT_EMPTYENCAPINFO);
			break;
		case 'a':
			if (encapinfo_parse_directive(optarg, &pkginfo,
						      errbuf,
						      sizeof(errbuf)) == -1)
			{
				fprintf(stderr,
					"mkencap: invalid encapinfo directive "
					"\"%s\": %s\n", optarg, errbuf);
				usage();
				/* NOTREACHED */
			}
			break;
		case 'h':
		default:
			usage();
			/* NOTREACHED */
		}
	}

	/* set verbosity for download and archive modules */
	dl_verbose = verbose;
	archive_verbose = verbose;

	/* make sure our arguments are sane */
	if (argc - optind != 1)
	{
		usage();
		/* NOTREACHED */
	}

	if (mode == 0)
		BIT_SET(mode, MODE_CREATE | MODE_ENCAPINFO);

	/*
	** validate Encap package format version
	** and set platform name (needed before profile parsing)
	*/
	if (encap_vercmp(pkginfo.ei_pkgfmt, "2.0") == 0)
	{
		if (platform[0] == '\0')
			encap20_platform_name(platform, sizeof(platform));
	}
	else if (encap_vercmp(pkginfo.ei_pkgfmt, "2.1") == 0)
	{
		if (platform[0] == '\0')
			encap_platform_name(platform, sizeof(platform));
		if (platform_suffix != NULL)
		{
			strlcat(platform, "-", sizeof(platform));
			strlcat(platform, platform_suffix, sizeof(platform));
		}
	}
	else
	{
		fprintf(stderr, "mkencap: unsupported Encap package format "
			"version \"%s\"\n", pkginfo.ei_pkgfmt);
		exit(1);
	}

	/* get current directory */
	if (getcwd(cwd, sizeof(cwd)) == NULL)
	{
		fprintf(stderr, "mkencap: getcwd(): %s\n", strerror(errno));
		exit(1);
	}

	/* initialize global paths */
	init_encap_paths(cwd, opt_src, opt_tgt);

	if (verbose > 1)
		printf("### source directory:\t\t%s\n", source);

	/*
	** set contact field here, because $ENCAP_CONTACT won't be
	** available when we need it in build.c
	*/
	if (BIT_ISSET(mode, MODE_BUILD|MODE_ENCAPINFO)
	    && !BIT_ISSET(mkencap_opts, MKENCAP_OPT_EMPTYENCAPINFO))
	{
		mk_set_encapinfo_contact(&pkginfo);

		/* FIXME */
		pkginfo.ei_platform = strdup(platform);
	}

#ifdef HAVE_LIBEXPAT
	/* in build mode, read package profile to get pkgspec */
	if (BIT_ISSET(mode, MODE_BUILD))
	{
		if (profile_modes == 0)
			BIT_SET(profile_modes, MKENCAP_PROFILE_DOWNLOAD
					       | MKENCAP_PROFILE_UNPACK
					       | MKENCAP_PROFILE_PATCH
					       | MKENCAP_PROFILE_CONFIGURE
					       | MKENCAP_PROFILE_BUILD);

		mkencap_init_paths(cwd, opt_download, opt_common_src,
				   opt_build);

		if (verbose > 1)
		{
			printf("### target directory:\t\t%s\n", target);
			printf("### download directory:\t\t%s\n",
			       download_tree);
			printf("### common source tree:\t\t%s\n",
			       (common_src_tree[0] != '\0'
			        ? common_src_tree
			        : "N/A"));
			printf("### build tree:\t\t\t%s\n", build_tree);
			putchar('\n');
		}

		if (argv[optind][0] == '/')
			strlcpy(profile, argv[optind], sizeof(profile));
		else
			snprintf(profile, sizeof(profile), "%s/%s",
				 cwd, argv[optind]);
		epp = parse_profile(profile, m4_cmd, opt_m4_outfile);
		if (epp == NULL)
			exit(1);

		/* install mode creates the encapinfo file for us */
		if (BIT_ISSET(profile_modes, MKENCAP_PROFILE_INSTALL))
			BIT_CLEAR(mode, MODE_ENCAPINFO);

		/*
		** don't honor "-a" options in build mode
		** FIXME: this is really messy... 
		*/
		encap_list_empty(pkginfo.ei_ex_l, free);
		encap_list_empty(pkginfo.ei_rl_l, free);
		encap_list_empty(pkginfo.ei_ld_l, free);
		encap_list_empty(pkginfo.ei_pr_l, free);
		encap_hash_empty(pkginfo.ei_ln_h, free);

		if (build_package(epp, profile_modes, profile,
				  opt_env_file) != 0)
		{
			exitval = 1;
			goto mkencap_done;
		}

		pkgspec = epp->ep_pkgspec;
	}
	else
#endif /* HAVE_LIBEXPAT */
	{
		pkgspec = argv[optind];
		if (pkgspec[strlen(pkgspec) - 1] == '/')
			pkgspec[strlen(pkgspec) - 1] = '\0';
	}

	/*
	** set package directory (used by both encapinfo and create modes)
	*/
	snprintf(pkgdir, sizeof(pkgdir), "%s/%s", source, pkgspec);

	/* write encapinfo file */
	if (BIT_ISSET(mode, MODE_ENCAPINFO))
	{
		if (verbose > 1)
			printf("### Encap package format:\t%s\n\n",
			       pkginfo.ei_pkgfmt);

		if (mk_encapinfo(pkgdir) != 0)
		{
			exitval = 1;
			goto mkencap_done;
		}
	}

	/* create a package archive */
	if (BIT_ISSET(mode, MODE_CREATE))
	{
		/* set default archive file name */
		if (outfile[0] == '\0')
		{
			/*
			** if not in build or encapinfo mode, read
			** existing encapinfo file to get platform and
			** encap format version
			*/
			if (!BIT_ISSET(mode, MODE_BUILD)
			    && !BIT_ISSET(mode, MODE_ENCAPINFO))
			{
				if (encap_open(&e, source, target, pkgspec,
					       OPT_DEFAULTS,
					       mkencap_print_func) == -1)
				{
					fprintf(stderr, "mkencap: "
						"encap_open(\"%s\"): %s\n",
						pkgspec, strerror(errno));
					exitval = 1;
					goto mkencap_done;
				}

				eip = &e->e_pkginfo;
			}
			else
				eip = &pkginfo;

			/*
			** set default output filename
			** based on Encap package format version
			*/
			if (encap_vercmp(eip->ei_pkgfmt, "2.0") == 0)
			{
				snprintf(outfile, sizeof(outfile),
					 "%s.tar.gz",
					 pkgspec);
			}
			else	/* must be Encap 2.1, since we checked above */
			{
				snprintf(outfile, sizeof(outfile),
					 "%s-encap-%s.tar.gz", pkgspec,
					 (eip->ei_platform != NULL
					  ? eip->ei_platform
					  : platform));
			}

			if (e != NULL)
				encap_close(e);
		}

		if (verbose > 1)
			printf("### output file:\t\t%s\n\n", outfile);

		i = ARCHIVE_OPT_ENCAPINFO;
		if (BIT_ISSET(mkencap_opts, MKENCAP_OPT_OVERWRITE))
			i |= ARCHIVE_OPT_FORCE;

		if (archive_create(outfile, pkgdir, i) != 0)
			exitval = 1;
	}

  mkencap_done:
	encapinfo_free(&pkginfo);
	download_cleanup();
	exit(exitval);
}


